src/main/java/org/elasticsearch/index/query/QueryParseContext.java

```diff
@@ -310,22 +309,34 @@ public Query parseInnerQuery() throws IOException, QueryParsingException {
         if (queryParser == null) {
             throw new QueryParsingException(index, "No query registered for [" + queryName + "]");
         }
-        Query result = queryParser.parse(this);
+        QueryBuilder result = queryParser.fromXContent(this);
         if (parser.currentToken() == XContentParser.Token.END_OBJECT || parser.currentToken() == XContentParser.Token.END_ARRAY) {
             // if we are at END_OBJECT, move to the next one...
             parser.nextToken();
         }
-        if (result instanceof NoCacheQuery) {
+        return result;
+    }
+
+    @Nullable
+    public Query parseInnerQuery() throws IOException, QueryParsingException {
```

Comment:
we might want to add a comment here or deprecate it given that we will need to use the above new method instead, once all of the queries are migrated?